repository_service_tuf_worker/repository.py

```diff
@@ -607,108 +679,388 @@ def save_settings(self, root: Metadata[Root], roles_info: Dict[str, Any]):
                 )
-                targets.signed.add_key(online_pub_key, role_name)
-                custom_target = Metadata(Targets())
-                self._bump_expiry(custom_target, role_name)
-                self._sign(custom_target)
-                self._persist(custom_target, role_name)
-                delegated_roles.append(role_name)
+            )
+            self._bump_expiry(bins_role, BINS)
+            self._sign(bins_role)
+            self._persist(bins_role, delegated_name)
+            delegated_roles.append(delegated_name)
+
+        targets_crud.create_roles(self._db, db_target_roles)
+
+    def _remove_delegated_role_keys(
+        self, targets: Metadata[Targets], delegated: DelegatedRole
+    ):
+        other_keys = []
+        logging.debug("mapping keys")
+        for not_delete in targets.signed.delegations.roles:
+            if not_delete == delegated:
+                continue
+            other_keys += targets.signed.delegations.roles[not_delete].keyids
+
+        role_keys = targets.signed.delegations.roles[delegated.name].keyids
+        for key in role_keys:
+            if key not in other_keys:
+                logging.debug(f"removing key id {key}")
+                targets.signed.delegations.keys.delete(key)
+            else:
+                logging.debug(f"key {key} used by other role")
+
+    def _add_delegated_role_keys(
+        self,
+        targets: Metadata[Targets],
+        delegations: Delegations,
+        role_name: str,
+        role_metadata: Optional[Metadata[Targets]] = None,
+    ):
+        if len(delegations.roles[role_name].keyids) == 0:
+            logging.debug(f"role '{role_name}' without key, adding online key")
+            delegations.roles[role_name].keyids.append(self._online_key.keyid)
+
+        for keyid in delegations.roles[role_name].keyids:
+            if keyid not in targets.signed.delegations.keys.keys():
+                logging.debug(f"added key id {keyid}")
+                if delegation_key := delegations.keys[keyid]:
+                    targets.signed.delegations.keys[keyid] = delegation_key
+                else:
+                    raise ValueError(f"role {role_name} has inconsistent keys")
+            if keyid == self._online_key.keyid and role_metadata:
+                logging.debug(f"role '{role_name}' using online key, signing")
+                self._sign(role_metadata)
+
+    def _update_delegated_roles(
+        self,
+        targets: Metadata[Targets],
+        updated_delegated: DelegatedRole,
+        old_delegated: DelegatedRole,
+    ):
+        # NOTE: we don't support renaming and we verify it before
+        role_name = updated_delegated.name
+        # we can detect if keys are different, remove and add
+        if updated_delegated.keyids != old_delegated.keyids:
+            logging.debug("Identified different keys remove and add")
+
+            # remove keys
+            self._remove_delegated_role_keys(targets, old_delegated)
```

Comment:
Why don't you add the new keys from `updated_delegated` after you have removed the old ones?

Reply:
We do it using the `Delegation`

https://github.com/repository-service-tuf/repository-service-tuf-worker/blob/ee61ed4499d090933e77fdff4a7a631686925b28/repository_service_tuf_worker/repository.py#L976-L978

Reply:
I wouldn't separate removing and adding keys in different functions.
Will add it to an issue.

Reply:
https://github.com/repository-service-tuf/repository-service-tuf-worker/issues/595